src/apitest/buffer_set_lines.c

```diff
@@ -150,6 +150,35 @@ MU_TEST(test_modified_is_set)
   char_u *lines[] = {"one"};
   vimBufferSetLines(curbuf, 0, 0, lines, 1);
   mu_check(vimBufferGetModified(curbuf) == TRUE);
+}
+ MU_TEST(set_lines_into_empty_buffer)
+{
+  buf_T *buf = vimBufferOpen(NULL, 0, 0);
+
+  char_u* lines[] = {
+    "line 1",
+    "line 2",
+    "line 3",
+    "line 4",
+    "line 5",
+    "line 6"
+  };
+
+  int i;
+
+  vimBufferSetLines(buf, 1, 6, lines, 6);
+
+  for (i = 0; i < 6; ++i) {
+    printf("LINE %i: %s\n", i, vimBufferGetLine(buf, i));
+  }
+
+  mu_check(strcmp(vimBufferGetLine(buf, 1), "line1") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 2), "line2") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 3), "line3") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 4), "line4") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 5), "line5") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 6), "line6") == 0);
+
```

Comment:
I think there is a few things that could be happening here, since this gets very confusing.

Vim itself, the part we all interact with as an editor, is all one based. The first line of your file is line 1 etc. So `vimBufferGetLine` should be using 1 based indexes, not 0 based ones in the `for` loop you've got there.

However, the actual internal representation is more programmatic so it is zero indexed, so you'd want to use 0 for the set lines. (That is at least how I see it, line numbers vs Array indices, I *think* its valid and holds up to think of it that way).

If you change the `vimBufferSetLines` to be zero indexed and to go from 0 to -1, does this test work as expected? I believe we use -1 to signify the full buffer should be updated, same as vim/neovim. Admittedly that is hard to know from just the code unless you happen to have read the vim help pages before hand. (You don't have to use -1, it just an easier way of specifying "the whole buffer").

You can also see how Oni2 uses this code, both in our C bindings https://github.com/onivim/reason-libvim/blob/master/src/bindings.c (Wrapping the C code to be used in our Reason Code) and then the reason code itself, which is usually just a straight call of the external C code, but sometimes has additional logic: https://github.com/onivim/reason-libvim/blob/master/src/Buffer.re. Both of those links are to the MIT licensed archived repo, I'm not 100% what license the version that is inline in the Oni2 repo has.

Reply:
The documentation comment for `vimBufferSetLines` says it is one based. At least, that is how I interpret it:

```
/*
 * vimBufferSetLines
 *
 * Set a range of lines from the one-based start line to one-based end, inclusive.
 * 
 * Examples:
 * vimBufferSetLine(buf, 1, 1, ["abc"]); // Set line 1 to "abc""
 * vimBufferSetLine(buf, 1, 2, ["abc"]); // Remove line 2, set line 1 to "abc"
 * vimBufferSetLine(buf, 0, 0, ["def"]); // Insert "def" before the contents of the buffer
 */
```

And that using zero means "append to the top of the file", or at least again how I interpreted it.

It's also unexpected that you would use `0` in set line and `1` in get line.

Reply:
Hm, it is possible that the doc comment is out of date or wrong, or that I'm misunderstanding the Reason Code in `Buffer.re`, since that defaults to `0` and `-1` if a start or end isn't given.

This patch works for me locally at least

```diff
--- a/src/apitest/buffer_set_lines.c
+++ b/src/apitest/buffer_set_lines.c
@@ -166,18 +166,18 @@ MU_TEST(test_modified_is_set)
 
   int i;
 
-  vimBufferSetLines(buf, 1, 6, lines, 6);
+  vimBufferSetLines(buf, 0, 6, lines, 6);
 
-  for (i = 0; i < 6; ++i) {
+  for (i = 1; i < 7; ++i) {
     printf("LINE %i: %s\n", i, vimBufferGetLine(buf, i));
   }
 
-  mu_check(strcmp(vimBufferGetLine(buf, 1), "line1") == 0);
-  mu_check(strcmp(vimBufferGetLine(buf, 2), "line2") == 0);
-  mu_check(strcmp(vimBufferGetLine(buf, 3), "line3") == 0);
-  mu_check(strcmp(vimBufferGetLine(buf, 4), "line4") == 0);
-  mu_check(strcmp(vimBufferGetLine(buf, 5), "line5") == 0);
-  mu_check(strcmp(vimBufferGetLine(buf, 6), "line6") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 1), "line 1") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 2), "line 2") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 3), "line 3") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 4), "line 4") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 5), "line 5") == 0);
+  mu_check(strcmp(vimBufferGetLine(buf, 6), "line 6") == 0);
 
 }
```

Additionally, the other tests in this file also seem to use a zero-indexed way for `set`, ie: 

```c
  char_u *lines[] = {"after first line"};
  vimBufferSetLines(curbuf, 1, 1, lines, 1);
```

but if you attempt to use `0` for get, you will get a duplicated first line.

So I guess it is out of date? There was a discussion back and forth on the best way to line these up I believe as its casued a bunch of confusion. The rest of the file and the wrapping code we use for Oni2 at least uses a 0 based index from my understanding.

@bryphe will hopefully have a better memory for the reasoning behind the difference, since it is confusing at first.

Reply:
Sounds good. I suspect the doc comment is out of date. FWIW, I'd vote make both zero based. 

This discussion was my real intent of this PR, not to get it merged. I figured it was easier to talk about this with an actual diff instead of just code snippets in an issue.

Reply:
I can change this PR to be a doc update instead. I'll wait to see what Bryan thinks.

Reply:
Just saw your tweet @city41 - integrated libvim to godot is an awesome idea!

> So I guess it is out of date? There was a discussion back and forth on the best way to line these up I believe as its casued a bunch of confusion. The rest of the file and the wrapping code we use for Oni2 at least uses a 0 based index from my understanding.

Yeah, it's definitely confusing! I used one-based here just because it was close to the internal representation of Vim, so it kept the wrapper thin. In Onivim 2 - we do some massaging on top of that representation - the C binding API is [here](https://github.com/onivim/reason-libvim/blob/dc4740a87833e0587eae03b17912e10aaaa6cfc6/src/bindings.c#L625) and the tweaking we do to the indices is [here](https://github.com/onivim/reason-libvim/blob/dc4740a87833e0587eae03b17912e10aaaa6cfc6/src/Buffer.re#L73). That wrapper API let's us optional specify the start/end - if neither are specified, the whole buffer gets replaced.

I'd be open to switching to 0-based indices - it might be a decent amount of work though, as there's some other APIs we stick to Vim's internal representation, ie - `vimSearchGetHighlights`, `vimUndoSaveRegion`, `vimCursorGetLine`, etc. 

In the meantime - a doc update sounds great! Having a fresh perspective on it is very helpful 👍

Reply:
> Just saw your tweet @city41 - integrated libvim to godot is an awesome idea!

Woot! Thanks to you all for making this lib. It makes adding vim to Godot so simple.

> In the meantime - a doc update sounds great! Having a fresh perspective on it is very helpful +1

Ok, one sec and I'll update this PR. And just to be clear, we all seem to agree that `vimBufferSetLines` is actually zero based, so its docs need a tweak.

Reply:
But what about this test? If `vimBufferSetLines` is zero based, then the outcome of this test is not what I'm expecting

```
MU_TEST(test_append_after_buffer)
{
  char_u *lines[] = {"after"};
  vimBufferSetLines(curbuf, 3, 4, lines, 1);

  printf("LINE 1: %s\n", vimBufferGetLine(curbuf, 1));
  printf("LINE 2: %s\n", vimBufferGetLine(curbuf, 2));
  printf("LINE 3: %s\n", vimBufferGetLine(curbuf, 3));
  printf("LINE 4: %s\n", vimBufferGetLine(curbuf, 4));
  mu_check(strcmp(vimBufferGetLine(curbuf, 4), "after") == 0);
  mu_check(strcmp(vimBufferGetLine(curbuf, 3), "This is the third line of a test file") == 0);

  mu_check(vimBufferGetLineCount(curbuf) == 4);
}
```

```
-- START test_append_after_buffer --
LINE 1: This is the first line of a test file
LINE 2: This is the second line of a test file
LINE 3: This is the third line of a test file
LINE 4: after
.
.
.

-- END test_append_after_buffer --
```

Why does `after` end up on line 4?

Reply:
> Woot! Thanks to you all for making this lib. It makes adding vim to Godot so simple.

I remember someone asking about this a looong time ago on our Discord - cool to see it happen! 🎉 

> Ok, one sec and I'll update this PR. And just to be clear, we all seem to agree that vimBufferSetLines is actually zero based, so its docs need a tweak.

I think the one confusion is that, the `vimBufferSetLines` treats zero as 'before' the first line:

- `vimBufferSetLines(buf, 0, 0, lines)` <-- insert lines _before_ the first `lines` 
- `vimBufferSetLines(buf, 0, 1, lines)` <-- replace the first line with `lines`
- `vimBufferSetLines(buf, 1, 1, lines)` <-- insert `lines` between the first and second line
- `vimBufferSetLines(buf, 1, 2, lines)` <-- replace the second line with `lines`

So you could consider both the start and end as zero-based indices, if you consider the start as `inclusive` and end as `exclusive`. It might be worth adding some of these examples too (there's a few in the docs right now) - might help solidify the understanding.

Reply:
Ahhhhh, that makes total sense. But hmmm, it's hard to explain. I'll ponder some more, and if you all have any ideas let me know.

Reply:
Ok, I think the new documentation makes more sense (I updated this PR twice). If you all agree, I'll leave it at that.